File: wip-tmp/progp21_simonqv-X2/master/Erlang/src/bernoulli.erl
-module(bernoulli).
-compile(export_all).
-import(lists, [append/2,nth/2]).
-import(math, [pow/2]).

main() -> b(20).


b(N) ->
  B = lists:append([], [1.0]),
  Bdone = outer_loop(1, N, B),
  printer(Bdone).


outer_loop(M, N, B) when M == N -> B;
outer_loop(M, N, B) ->
  Bm = (inner_loop(M-1, 0, B, 0.0)) / (M+1),
  Pow = pow(10, -8),
  X = if
    abs(Bm) < Pow -> 0.0;
    abs(Bm) >= Pow -> Bm
  end,
  outer_loop(M+1, N, lists:append(B, [X])).

inner_loop(M, K, [], Bm) when K == M-1 -> Bm;
inner_loop(M,K, [], Bm) when K /= M-1 -> Bm;
inner_loop(M, K, [B|BT], Bm) ->
  inner_loop(M, K+1, BT, (Bm - (binom(M+2, K) * B))).




binom(N, K) -> binom_rec(N,K,1,1).

binom_rec(_,K,I,R) when K == I - 1 -> R;
binom_rec(N,K,I,R) ->
  binom_rec(N,K,(I+1), ((R*(N-I+1))/(I))).

printer([B|[]]) -> io:fwrite("~*.*.0f~n", [9, 6, B]);
printer([B|BT]) ->
  io:fwrite("~*.*.0f~n", [9, 6, B]),
  printer(BT).

File: wip-tmp/progp21_simonqv-X2/master/Rust/Bernoulli.rs
fn main() {
	let size = 20;
    let vec = bernoulli(size);
    for i in 0..(size - 1) {
    	let val = vec.get(i as usize).unwrap();
    	
    	if (val.abs() as f64) < ((f64::powf(10.0,-8.0)) as f64) {
    		println!("{}", 0.0);
    	} else {
    		println!("{}", val);	
    	}
    }
}


fn bernoulli(n: u32) -> Vec<f64> {
	let mut vec = Vec::new();
	vec.push(1.0);
	for m in 1..n {
		let mut bm = 0.0;
		for k in 0..(m) {
			bm = bm - (binom(m+1, k) * vec.get(k as usize).unwrap());
		}
		bm = bm / (m as f64 + 1.0);
		vec.push(bm);
	}
	return vec;
}

fn binom(n: u32, k: u32) -> f64 {
	let mut r = 1.0;
	for i in 1..(k+1) {
		r = r * (n as f64 - i as f64 + 1.0) / i as f64;
	}
	return r;
}